### PR TITLE
#41 table relative paths

### LINKED ISSUE #41: Relative paths in remote datapackages don't work
# Overview

```javascript
const Datapackage = require('datapackage-test').Datapackage
const descriptor = 'https://raw.githubusercontent.com/frictionlessdata/datapackage-js/master/data/dp1/datapackage.json'

new Datapackage(descriptor).then(dp => {
    dp.resources[0].table.then(table => console.log(table.read()))
})
```

```
Error: ENOENT: no such file or directory, open 'data.csv'
```

# Tasks

- [ ] fix broken behavior
- [ ] add tests

### LINKED ISSUE #52: Buggy resource path resolution
https://github.com/frictionlessdata/datapackage-js/blob/master/src/resource.js#L73

```
  get source() {
    if (this._sourceKey === 'path' && this._basePath) {
      const resourcePath = this.descriptor[this._sourceKey]
      return path.normalize(`${this._basePath}/${resourcePath}`)
    }

    return this.descriptor[this._sourceKey]
  }
```

This is not the right logic: we need to check whether resource `path` attribute is a fully qualified url or not before prepending the basepath - we only prepend base if `path` is not a fully qualified url.

Aside: as general code practice i think this should be an if / else statement not if then a return - the logic flow would be clearer.

Aside: `type` as a function seems a too generic name -- typeOfResourcePath would be a more informative name and would obviate much of the documentation ;-)

### PR DESCRIPTION
**Pull request by dumyan**
- fixes #41 
- fixes #52

---

Resolves #41 by adding `basePath` for remote datapackages and after that using `url.resolve` in `Resources` class to point to the relative resource.

### CODE REVIEW

File: test/datapackage.js [hunk 1]
```diff
@@ -185,13 +185,23 @@ describe('Datapackage', () => {
...
-      const table2 = await datapackage.resources[0].table
-      const data = await table2.read()
+      const table = await datapackage.resources[0].table
+      const data = await table.read()
+
+      assert(_.isEqual(data, [['gb', 100], ['us', 200], ['cn', 300]]), 'Invalid data.')
+    })
+
+    it.only('URL resource', async () => {
+      const descriptor = 'https://dev.keitaro.info/dpkjs/datapackage.json'
```
Comment (dumyan) [inline test/datapackage.js]: The URL is temporary, once this is merged this will be changed to the newly added `data/dp4-remote-resource/datapackage.json`

File: test/datapackage.js [hunk 2]
```diff
@@ -185,6 +186,40 @@ describe('Datapackage', () => {
...
+
+      const datapackage = await new Datapackage(descriptor)
+      const table = await datapackage.resources[0].table
+      const data = await table.read()
+
+      assert(_.isEqual(data, [['gb', 100], ['us', 200], ['cn', 300]]), 'Invalid data.')
+    })
+
+    it('URL resource', async () => {
+      const descriptor = 'https://dev.keitaro.info/dpkjs/datapackage.json'
```
Comment (dumyan) [inline test/datapackage.js]: This is temporary, we will change it to github once this branch is merged.

File: src/resource.js [hunk 3]
```diff
@@ -71,12 +72,19 @@ export default class Resource {
...
+    const source = this.descriptor[this._sourceKey]
+
+    if (this._sourceKey === 'path' && this._basePath
+        && !Utils.isRemoteURL(source)) {
+
+      if (Utils.isRemoteURL(this._basePath)) {
+        return url.resolve(this._basePath, source)
+      }
+
+      return path.normalize(`${this._basePath}/${source}`)
```
Comment (roll) [inline src/resource.js]: I suppose we better use something like path join instead of string interpolation to make it cross-platform.

File: src/resource.js [hunk 4]
```diff
@@ -71,12 +72,19 @@ export default class Resource {
    * @returns {String|Array|Object}
    */
   get source() {
```
Comment (roll) [inline src/resource.js]: This function logic seems a little bit complicated - may be it's possible to make it more straightforward using comments and groups? We also should analyze it for vulnerabilities (accessing e.g. `/etc/passwd` using relative paths) so this code should be pretty clear to understand. 

For example:
```javascript
// Inline data
if (this.type === 'inline') return source;

// Local data
...

// Remote data
...
```

Comment (Fak3) [inline src/resource.js]: As for relative paths - it should be addresed on the spec level first. It already has a restriction of usage of dot-paths, but it is not enough (forbids usage only in the start of that path)

Comment (roll) [inline src/resource.js]: @Fak3 
One of the spec WG meetings has decided to ban any traverses outside of `datapackage.json` basedir for local paths. So we consider it as a rule just waiting official finalization.

**Comment by roll** [comment]
@dumyan 
One other thing - I suppose we need much more tests for path resolutions. Could you write a few additional tests (for datapackage and resource) with comprehensive list of combinations (basePath/path) to cover all edge cases?

```javascript
it('source/type correctly works with base path', async () => {
  const paths = [
    {base: 'folder1', path: 'data/data.csv', source: 'folder1/data/data.csv', type: 'local'},
    // ... like 10 more)
  ]
  paths.forEach({base, path, source, type} => {
    const resource = new Resource({path}, base)
    assert(resource.source === source)
    assert(resource.type === type)
  })
})
```

The same technique could be used for other tests to ensure `Resource.source/type` works correctly on all possible input data.

File: src/resource.js (hunk 4 above)
Comment (dumyan) [inline src/resource.js]: @Fak3 @roll 

both comments have been addressed. Path checking searches for `^/`, and occurences of `..` or `.\.` etc (backslash is being ignored when travesting dirs, so it's first taken out of the path).

File: src/resource.js [hunk 5]
```diff
@@ -46,15 +45,14 @@ export default class Resource {
    *
    * @returns {String}
    */
-  get type() {
+  get typeOfResourcePath() {
```
Comment (roll) [inline src/resource.js]: Please follow https://github.com/frictionlessdata/stack/issues/2 (`Resource.type` for now)

File: src/resource.js [hunk 6]
```diff
@@ -71,12 +69,36 @@ export default class Resource {
...
+      // Neither inline data nor path available
+      return null
+
+    } else if (this.typeOfResourcePath === 'inline') {
+      // Data is inline
+      return source
 
-    return this.descriptor[this._sourceKey]
+    } else if (this._sourceKey === 'path' && this._basePath === '') {
+      // Local path, no basePath provided
```
Comment (roll) [inline src/resource.js]: Not sure why `this.__sourceKey === 'path'` means that a local path:

```javascript
  get _sourceKey() {
    const inlineData = this.descriptor.data

    if (inlineData) {
      return 'data'
    }

    return 'path'
  }
```

Based on this code remote path also satisfies this condition.

File: src/resource.js [hunk 7]
```diff
@@ -71,12 +69,36 @@ export default class Resource {
...
+      if (this._validPaths) {
+        return source
+      }
+    } else if (this._sourceKey === 'path' && this._basePath !== '') {
+      // basePath needs to be prepended
+
+      // we need to check the validity of the paths here because one can use
+      // only the Resource class to read file contents with `table`
+      if (this._validPaths) {
```
Comment (roll) [inline src/resource.js]: See `basePath` comment in constructor

Comment (dumyan) [inline src/resource.js]: @roll 

Error is always thrown here if the paths are invalid.
